### ISSUE TITLE
v 0.42b does not check that NC is up to date before installing latest version

### ISSUE THREAD

**Issue by alento-group** [mail-in-a-box/mailinabox#1623]
v 0.42b does not check to see that NC has already been upgraded to v 14 which causes the backup restore to fail as it installs v 15 without upgrading the NC data structure.

**Comment by jvolkenant** [comment]
There is a lot to improve with MIAB.  If one follows the guide at https://mailinabox.email/maintenance.html to do a restore, the process is as follows (I believe):

- Do Fresh Install
- Install Latest MIAB
- run Duplicity restore (although, you should really shutdown a bunch of services before doing this (not mentioned in the guide)
- Rerun mailinabox (setup)


This causes problems because the version of nextcloud in /usr/local/lib/owncloud will differ than the users restored version in /home/user-data/owncloud

https://github.com/mail-in-a-box/mailinabox/blob/e37768ca8680106b037822975445de88e3e0da14/setup/nextcloud.sh#L96-L97
In the example of  the bullet points above, /usr/local/lib/owncloud will exist and /usr/local/lib/owncloud/version.php will contain the latest version, skipping the backup and "grep -q "OC_VersionString..." checks, yet the restored version (/home/user-data/owncloud) could be older.

That brings up an interesting point, what do we check then?
It looks like config.php a symlink, has a version line. (although doesn't exactly match $nextcloud_ver).  Perhaps there is another place/better place to check for the version of nextcloud in /home/user-data/owncloud versus the version in /usr/local/lib/owncloud

**Comment by jvolkenant** [comment]
On the topic of improvements to setup/nextcloud.sh:

- Currently it cannot update the version of apps without a new version of Nextcloud needing to be updated.
- On update, /usr/local/lib/owncloud is removed before downloading/verifying the new version.  Bad download = non functional Nextcloud
- There are different recommended app version for contact/calendar/external_apps for 13,14,15, but only one is being installed for every InstallNextcloud run.
- Is legacy "# ownCloud 8.1.1 broke upgrades"... code, still needed?
- Backup only backs up user data, does not backup app. Should have a better way to put back /home/user-data/owncloud, /usr/local/lib/owncloud and go back to old version if install fails.

Perhaps others.

**Comment by alento-group** [comment]
Well, I do not know what/why/how it happened ... but before Ubuntu 14 went EoL I updated my Ubuntu VPS's and ran the MiaB upgrade to v 0.30.  I had one MiaB running v 0.23 and another running v 0.18 or something crazy like that .... (bad, I know) .... anyways v0.30 perfectly checked the NC installs (Owncloud on the one I think) and upgraded them version by version ... so I assumed that there must have been some kind of check in v0.30 which may have not been implemented in v 0.42b .... /shrug

**Comment by alento-group** [comment]
https://discourse.mailinabox.email/t/solved-workaround-for-upgrading-over-multiple-versions-of-miab-to-the-latest-when-multiple-major-versions-of-nexcloud-are-included-first-accured-in-upgrade-from-v0-30-v0-40-nc13-to-v0-42b-nc15-with-v0-41-nc14-between/5233/34

**Comment by jvolkenant** [comment]
@alento-group correct.  Currently the nextcloud version detection is in /usr/local/lib/owncloud/version.php.  So the linked post is going about the only way currently available to restore a backup between versions. 

We will need to know the version of data in /home/user-data/owncloud in order to then know which versions to upgrade through.  I haven't looked into it, but unless there is a version string in something in /home/user-data/owncloud, or in the owncloud.db, we can  track one in a file, but that would only work for +v0.42b.

**Comment by JoshData** [comment]
It took me about an hour of searching, but it turns out `$STORAGE_ROOT/owncloud/config.php` contains the version. This can be accessed using:

    VERSION=$(php -r "include(\"$STORAGE_ROOT/owncloud/config.php\"); echo(\$CONFIG['version']);")

which gets the value of the version key.

This should solve the problem.

**Comment by anoma** [comment]
I have just tried to upgrade using the recommended procedure from 0.30 to 0.42b here https://discourse.mailinabox.email/t/mail-in-a-box-version-v0-40-and-moving-to-ubuntu-18-04/4289

I strongly recommend this page be given a warning this now as it simply doesn't work due to this issue.

What even more horrible is Nextcloud blocks MIAB actual mail services coming up and the installer fails to complete. 

After quite a bit of effort I have managed to get into a state where 0.41 is installed and mail works but Nextcloud is completely broken.

Now, correct me if I am wrong but I dont believe I use Nextcloud at all anyway. Unless something happens under the hood then would a viable emergency option be documenting how to reset ONLY Nextcloud to default after duplicity restore has taken place. 

If I am correct in my assumption this procedure would make the upgrade from 0.30 viable long term at the cost of reseting calendar and contacts. 

The problem is I dont actually know how to do this Nextcloud only reset. Anyone?

**Comment by rfrush** [comment]
Anoma, Josh-

Perhaps a link on the main 0.30->0.4x upgrade page should point to this thread (link below), which discusses how to manage the upgrade to 0.42b.   I used the process a few days ago to upgrade my server, and it worked without a hitch!

https://discourse.mailinabox.email/t/solved-workaround-for-upgrading-over-multiple-versions-of-miab-to-the-latest-when-multiple-major-versions-of-nexcloud-are-included-first-accured-in-upgrade-from-v0-30-v0-40-nc13-to-v0-42b-nc15-with-v0-41-nc14-between-updated-cose-of-backup/5233/32?u=just4t

**Comment by anoma** [comment]
This thread is what I found and followed with less success than you... albeit I have no idea if Nextcloud was working to start with. No one on this server ever uses it seems.

Given that, I would prefer to stick to the system as-is in a working state rather than have to revert and lose mail.

Being able to reset just Nextcloud to default seems like a sane thing to be able to do for more reasons other than upgrade. My gut is its probably quite easy for someone who know how MIAB bolts together.

**Comment by alento-group** [comment]
I am confused @anoma because you said that you followed the instructions in the "Moving to Ubuntu 18" thread originally ...
Or did you use the upgrade path in the "Workaround" thread? which you just stated in the last post that you had followed.
It would be helpful to know EXACTLY what you did so that the correct warning is issued.  Thanks!

**Comment by anoma** [comment]
I followed the official upgrade guide at first and when that failed, following a debug period and fix attempts of my own, I finally located and attempted to follow [the link you posted](https://discourse.mailinabox.email/t/solved-workaround-for-upgrading-over-multiple-versions-of-miab-to-the-latest-when-multiple-major-versions-of-nexcloud-are-included-first-accured-in-upgrade-from-v0-30-v0-40-nc13-to-v0-42b-nc15-with-v0-41-nc14-between-updated-cose-of-backup/5233/32?u=just4t)

The end result was some sort of in Frankenstein state where everything appears to work except for Nextcloud.

At this point I had two decisions to make:

- Could I handle a further outage of an unknown length to fix a subsystem that isn't actually needed
- Could I accept the loss in email this would result in.

Keep in mind that for a large percentage of the upgrade following the official guide mail was working and because of this timeline there is no way to reliably make a new backup that would fit in with either procedure type and result in no lost mail.

**Comment by alento-group** [comment]
@anoma.  Gotcha!  That was what I had suspected had happened but wanted to know rather than speculate.  Oh and please give @Wojtek the credit for that amazing amount of research and effort he put in to discovering that solution.  All I did was offer a bit of advice.

I do not know much concerning the inner workings of NC and MiaB in tandem.  Perhaps @joshdata and/or @jvolkenant can weigh in with some more information.  My personal opinion was always along the line of I do not use NC why is it even there?  Sadly many people who do use calendar/contacts did not much care for my attitude on the subject.

Now, how to proceed ... one thing that you have mentioned is downtime.  I just want to be sure that you are aware that properly configured mail servers (most) will retry emails if a mail server does not reply for up to 2-3 days, so mail loss should not be a factor in a decision not to take the server down for a short period of time to do maintenance/make repairs.

If the maintainers can advise more specifically what the role of NC is and how interwoven it is into MiaB, that would be helpful to determine how to proceed.  For like you, if NC got mangled, I would be perfectly content to leave it in a damaged state -- but would fear what other consequences that decision could bring down the road.

**Comment by api4user** [comment]
@alento-group I am with you on this one and do not understand why NextCloud is being used in this project.  If its use was suggested today there is no way @JoshData would allow it.

It seems NextCloud brings users nothing but trouble as it is not very stable (or updated too regularly) whereas something like [Radicale](https://radicale.org/) is very stable and would not add any cruft to mailinabox and only adds items directly related to mailinabox.

I keep meaning to remove NextCloud from my server and replace it with Radicale but I would see no point in submitting a pull request as it would just be a waste of time knowing @JoshData would never approve it.

**Comment by anoma** [comment]
Worth pointing out that it appears that if Nextcloud is in a damaged state mailinabox itself cannot be upgraded safely.

On one occasion, attempting the upgrade left mailinabox completely inoperable. I believe it was DNS that failed but that detail is somewhat irrelevant as i dont believe Nextcloud failing should ever cause the whole updater to bomb out.

**Comment by jvolkenant** [comment]
While this is a lot of off topic discussion.  I just want to make sure that I understand what problem is raised in this issue.  I want to make sure I can make a reproducible test of this issue.

A backup from v0.30 was restored to a new Ubuntu 18.04 install running a fresh install of Mail-in-a-Box v0.42b and Nextcloud was not upgraded from 13 => 14 => 15?

**Comment by anoma** [comment]
Correct.

If you follow the official procedure to upgrade from v0.30 to v0.40 (v0.42b now):

https://discourse.mailinabox.email/t/mail-in-a-box-version-v0-40-and-moving-to-ubuntu-18-04/4289

Nextcloud upgrades fail somewhere in the process and in doing so bomb out the entire installer resulting in a completely inoperable mailinabox install. No mail can be sent or received or websites seen.

**Comment by jvolkenant** [comment]
Appears that the ppa maintainer no longer offers 14.04 packages for [php7](http://ppa.launchpad.net/ondrej/php/ubuntu/pool/main/p/php7.0/) which makes doing a new install on 14.04 and MIAB v0.30 rather impossible.

I'll try from 18.04, MIAB v0.40, since that is still Nextcloud 13.

**Comment by anoma** [comment]
It has been an interesting discussion but I would not expend much effort on making the upgrade work as-is.

As detailed by yourself the fundamental architecture of the upgrade vs the backup makes it fragile at its very core.

All we need is a way to reset/blank Nextcloud content restored from backup and with that we have solved both the upgrade problem we have now and we also keep the door open for future update fixes well beyond anything we would officially support and would want to help straglers.

If no one comes forward with what needs deleted I will try and reverse engineer some kludge from a new install.

**Comment by jvolkenant** [comment]
@anoma,  While I do understand there are some improvements that can be made to MIAB and nextcloud.sh as a whole, I don't believe it is 'fragile at it's very core'

The issue raised here is upgrading through multiple versions of Nextcloud.  Currently nextcloud.sh checks the installed application for what version nextcloud is running, not what version the owncloud.db is. (Please see my posts above)  The problem will still happen down the line anytime someone skips intermediate versions of nextcloud.

Using the logic @JoshData posted above we can use proper version of nextcloud to use as our basis for doing upgrades.

While MIAB is mostly in a maintenance release (and has been for a while), we do accept patches to fix issues that arise such as this.  New features probably won't be accepted since that would add to the complexity of maintaining MIAB, but I would like to think this is a problem we can solve.

**Comment by JoshData** [comment]
The biggest mistake I made in this project was adding ownCloud (which became Nextcloud), and we've been through this in the discussion about replacing it in #514. I would love to drop it so long as there is a reasonable migration path for current users who are using it in supported ways.

If someone can implement and test the fix I suggested above, we can probably get this issue fixed tho.

Another path might be to have people migrate from v0.30 to v0.40 before going to the latest version. That should be possible with:

    curl -s https://mailinabox.email/setup.sh | sudo TAG=v0.40 bash

> If the maintainers can advise more specifically what the role of NC is and how interwoven it is into MiaB, that would be helpful to determine how to proceed.

Nextcloud isn't deeply interwoven. If it's broken, everything else should continue working fine. Removing `source setup/nextcloud.sh` from `setup/start.sh` should be fine to complete an install/upgrade without Nextcloud.

**Comment by anoma** [comment]
I can test bypassing Nextcloud in setup using your suggestion above in 48 hours time.

What should I delete to make the installer think Nextcloud needs new defaults. i.e reset just Nextcloud

**Comment by anoma** [comment]
Not that there was any doubt but commenting out:

```
# Start service configuration.
...
#source setup/nextcloud.sh
```

Allows all other aspects of the setup to complete and as far as I can tell the v0.30 to v0.42b restore from backup works perfectly.

**Comment by alento-group** [comment]
> Not that there was any doubt but commenting out:
> 
> ```
> # Start service configuration.
> ...
> #source setup/nextcloud.sh
> ```
> 
> Allows all other aspects of the setup to complete and as far as I can tell the v0.30 to v0.42b restore from backup works perfectly.

So, this is a bit off-topic ... but I appreciate the effort and want to clarify ... installing v0.42b with the above commented out will install without NextCloud ... and the backup restoration went without a hitch?  I would presume it would then be ok to delete the 'owncloud' related directories in /home/user-data/ once this is done, correct? :)  And again thanks!

**Comment by anoma** [comment]
My issue, like everyone else, is that the backup taken in v0.30 is not compatible with the NextCloud version included in v0.42b.

So there are two issues:

1. The installer fails to successlly upgrade NextCloud so you end up with MIAB v0.42b but NextCloud is broken
2. The failure to upgrade NextCloud  causes "sometimes" the entire Mailinabox installer to stop resulting in a partial install and no mail

There are currently 3 workarounds:

1. Disable the NextCloud portion of the update as listed above
2. Follow an unofficial upgrade path stepping through older versions of Mailinabox upgrading NextCloud in smaller supported steps (https://discourse.mailinabox.email/t/solved-workaround-for-upgrading-over-multiple-versions-of-miab-to-the-latest-when-multiple-major-versions-of-nexcloud-are-included-first-accured-in-upgrade-from-v0-30-v0-40-nc13-to-v0-42b-nc15-with-v0-41-nc14-between-updated-cose-of-backup/5233/32?u=just4t)
3. Wait for or patch with v0.42b https://github.com/mail-in-a-box/mailinabox/pull/1632

I would recommend option 3 as it also has some notes in the comments on how I fixed a failed install.

I do not know if you can delete 'owncloud' related directories in /home/user-data/ `. I would like to know how to reset NextCloud to defaults correctly but as off yet I havent found anything that says how to in Mailinabox

**Comment by alento-group** [comment]
Yes, option #3 sounds ideal.  I hope that @joshdata will release it as v0.42c ASAP!  :)

Thanks to everyone who worked so hard on this.

**Comment by remobi** [comment]
Just this note:
> Another path might be to have people migrate from v0.30 to v0.40 before going to the latest version. That should be possible with:
> 
> ```
> curl -s https://mailinabox.email/setup.sh | sudo TAG=v0.40 bash
> ```

Yes migrating from v0.30 to v0.40, then v0.41, then v0.42b (latest release ; NextCloud v15.0.8) worked for me today with:
step 1. ```curl -s https://mailinabox.email/setup.sh | sudo TAG=v0.40 bash```
step 2. ```curl -s https://mailinabox.email/setup.sh | sudo TAG=v0.41 bash```
step 3. ```curl -s https://mailinabox.email/setup.sh | sudo bash```

**Comment by jvolkenant** [comment]
Issue should be resolved with the merge of #1632
